src/index/hgraph_index.cpp

```diff
@@ -0,0 +1,675 @@
+        ids[j] = results.top().second;
+        results.pop();
+    }
+    return std::move(result);
+}
+
+tl::expected<BinarySet, Error>
+HGraphIndex::serialize() const {
+    if (GetNumElements() == 0) {
+        return empty_binaryset();
+    }
+    SlowTaskTimer t("hgraph serialize");
+    size_t num_bytes = this->cal_serialize_size();
+    try {
+        std::shared_ptr<int8_t[]> bin(new int8_t[num_bytes]);
+        auto buffer = reinterpret_cast<char*>(const_cast<int8_t*>(bin.get()));
+        BufferStreamWriter writer(buffer);
+        this->serialize(writer);
+        Binary b{
+            .data = bin,
+            .size = num_bytes,
+        };
+        BinarySet bs;
+        bs.Set(INDEX_HGRAPH, b);
+
+        return bs;
+    } catch (const std::bad_alloc& e) {
+        LOG_ERROR_AND_RETURNS(
+            ErrorType::NO_ENOUGH_MEMORY, "failed to serialize(bad alloc): ", e.what());
+    }
+}
+
+tl::expected<void, Error>
+HGraphIndex::deserialize(const ReaderSet& reader_set) {
+    SlowTaskTimer t("hgraph deserialize");
+    if (this->GetNumElements() > 0) {
+        LOG_ERROR_AND_RETURNS(ErrorType::INDEX_NOT_EMPTY,
+                              "failed to deserialize: index is not empty");
+    }
+
+    try {
+        auto func = [&](uint64_t offset, uint64_t len, void* dest) -> void {
+            reader_set.Get(INDEX_HGRAPH)->Read(offset, len, dest);
+        };
+        uint64_t cursor = 0;
+        auto reader = ReadFuncStreamReader(func, cursor);
+        this->deserialize(reader);
+    } catch (const std::runtime_error& e) {
+        LOG_ERROR_AND_RETURNS(ErrorType::READ_ERROR, "failed to deserialize: ", e.what());
+    }
+
+    return {};
+};
+
+void
+HGraphIndex::hnsw_add(const DatasetPtr& data) {
+    uint64_t total = data->GetNumElements();
+    auto* ids = data->GetIds();
+    auto* datas = data->GetFloat32Vectors();
+    auto cur_count = this->bottom_graph_->TotalCount();
+    vsag::Vector<std::shared_mutex>(total + cur_count, allocator_).swap(this->neighbors_mutex_);
+
+    auto build_func = [&](InnerIdType begin, InnerIdType end) -> void {
+        for (InnerIdType i = begin; i < end; ++i) {
+            int level = this->get_random_level() - 1;
+            auto label = ids[i];
+            auto inner_id = i + cur_count;
+            {
+                std::unique_lock<std::shared_mutex> lock(this->label_lookup_mutex_);
+                this->label_lookup_[label] = inner_id;
+            }
+
+            std::unique_lock<std::mutex> lock(this->global_mutex_);
+            bool need_lock = false;
+            if (level >= int64_t(this->max_level_) || bottom_graph_->TotalCount() == 0) {
+                for (int64_t j = max_level_; j <= level; ++j) {
+                    this->route_graphs_.emplace_back(this->generate_one_route_graph());
+                }
+                max_level_ = level + 1;
+                need_lock = true;
+            } else {
+                lock.unlock();
```

Comment:
Don't need a lock most of the time?

Reply:
yes, the lock only for some global members not for detail graph nodes and neighbors